### DISCUSSION TITLE
Placing a figure at the top of page and changing it with section

### DISCUSSION THREAD

**Discussion by psads-git** [typst/typst#4464] [Q&A] [answered]
Consider the following code. I would like to place the rectangle at the top of each page of section 1 and the circle at the top of each page of section 2. Is that possible? Thanks!

```
#rect(width: 20mm, height:  10mm)

= Section 1

#lorem(10000)

#circle(radius: 5mm)

== Section 2

#lorem(5000)
```

**Comment by Mc-Zen** [comment, marked as answer]
Hi, you are probably looking for a page `header`. 

You can achieve this with the following code. I will explain it below. 
```typ
#set heading(numbering: "1")

#let get-current-heading() = {
  let loc = here()
  let heading-selector = heading.where(level: 1, outlined: true)
  let el = query(heading-selector.after(loc)).at(0, default: none)
  if el != none and el.location().page() == loc.page() {
    return el
  } else {
    return query(heading-selector.before(loc)).at(-1, default: none)
  }
}

#pagebreak() // Let's test that nothing is shown before the first section

#set page(header: context {
  // let chapter-num = counter(heading).at(here()).first() // Does not suffice
  let chapter-num = counter(heading).at(get-current-heading().location()).first()
  if chapter-num == 1 {
    rect(width: 20mm, height:  10mm)
  } else if chapter-num == 2 {
    circle(radius: 5mm)
  }
})

= First Section

#lorem(1500)

// #pagebreak()
= Second Section

#lorem(1000)

```

The main part is in the middle, where we call `set page(header: ...)`. The content of the header may be dependent of the page that the header appears on. Specifically, we want to know the state of the `heading` counter on the page. Let us first look at how this can be done in principle (but this will not be enough here, see below):
```
let chapter-num = counter(heading).at(here()).first() // Does not suffice
```
We query the counter value for headings at the current location (`here`) and take the first value (the top-level section number). Unfortunately, `here` returns the current location which - for the header - is the top of the page. At the top, a potential new section on the page "is not yet there" which is not quite what you want since the symbol will then not appear (or the wrong symbol will show) on the _first_ page of the section. 

To fix it, we query for the first heading on the page we're on or - if there is none - the previous heading. From this heading we can finally obtain the location again and the state of the `heading` counter. 

Let me know if this did it for you

**Comment by Mc-Zen** [comment]
Additional note, there is also the package [hydra](https://typst.app/universe/package/hydra) that helps with creating headers. 

```typ
#import "@preview/hydra:0.4.0": hydra
#set page(header: context {
  hydra(heading,display: (_, el) => {
    let chapter-num = counter(heading).at(el.location()).first() // Does not suffice
    if chapter-num == 1 {
      rect(width: 20mm, height:  10mm)
    } else if chapter-num == 2 {
      circle(radius: 5mm)
    }
  }, skip-starting: false)
})
```

Remark: If your second section does not starts at the beginning of the page, it will still show the rectangle on this page (contrary to the other approach above). Depends on how you want it :)

**Comment by psads-git** [comment]
I have learned so much from your answers, @Mc-Zen! Thank you one million!

**Reply by Mc-Zen** [reply]
You are welcome!
